cql3/expr/expression.cc

```diff
@@ -216,21 +216,95 @@ get_value(const subscript& s, const evaluation_inputs& inputs) {
+// It could be represented by std::optional<bool>, but optional
+// can be implicitly casted to bool, which might cause mistakes.
+// (bool)(std::make_optional<bool>(false)) will return true,
+// despite the fact that the represented value is `false`.
+// To avoid any such problems this class is introduced
+// along with the is_true() method, which can be used
+// to check if the value held is indeed `true`.
+class bool_or_null {
+    std::optional<bool> value;
+public:
+    bool_or_null(bool val) : value(std::move(val)) {}
+    bool_or_null(null_value) : value(std::nullopt) {}
+
+    static bool_or_null null() {
+        return bool_or_null(null_value{});
+    }
+    bool has_value() const {
+        return value.has_value();
+    }
+    bool is_null() const {
+        return !has_value();
+    }
+    const bool& get_value() const {
+        return *value;
+    }
+    const bool is_true() const {
+        return has_value() && get_value();
+    }
+};
+
 /// True iff lhs's value equals rhs.
-bool equal(const expression& lhs, const managed_bytes_opt& rhs, const evaluation_inputs& inputs) {
-    if (!rhs) {
-        return false;
+bool_or_null equal(const expression& lhs, const managed_bytes_opt& rhs_bytes, const evaluation_inputs& inputs) {
+    raw_value lhs_value = evaluate(lhs, inputs);
+    if (lhs_value.is_unset_value()) {
+        throw exceptions::invalid_request_exception("UNSET_VALUE found on left-hand side of an equality operator");
     }
-    const auto value = evaluate(lhs, inputs).to_managed_bytes_opt();
-    if (!value) {
-        return false;
+    if (lhs_value.is_null() || !rhs_bytes.has_value()) {
+        return bool_or_null::null();
+    }
+    managed_bytes lhs_bytes = std::move(lhs_value).to_managed_bytes();
+
+    return type_of(lhs)->equal(managed_bytes_view(lhs_bytes), managed_bytes_view(*rhs_bytes));
+}
+
+static std::optional<std::pair<managed_bytes, managed_bytes>> evaluate_binop_sides(const expression& lhs,
+                                                                                   const expression& rhs,
+                                                                                   const oper_t op,
+                                                                                   const evaluation_inputs& inputs) {
+    raw_value lhs_value = evaluate(lhs, inputs);
+    raw_value rhs_value = evaluate(rhs, inputs);
+
+    if (lhs_value.is_unset_value()) {
+        throw exceptions::invalid_request_exception(
+            format("UNSET_VALUE found on left-hand side of a binary operator with operation {}", op));
     }
-    return type_of(lhs)->equal(managed_bytes_view(*value), managed_bytes_view(*rhs));
+    if (rhs_value.is_unset_value()) {
+        throw exceptions::invalid_request_exception(
+            format("UNSET_VALUE found on right-hand side of a binary operator with operation {}", op));
+    }
+    if (lhs_value.is_null() || rhs_value.is_null()) {
+        return std::nullopt;
```

Comment:
Wait, what if the operation is IS NOT NULL? Don't we need to return TRUE for that, not NULL?